crates/stark-backend/src/prover/hal.rs

```diff
@@ -0,0 +1,180 @@
+//! # Hardware Abstraction Layer
+//!
+//! Not all hardware implementations need to implement this.
+//! A pure external device implementation can just implement the [Prover](super::Prover) trait directly.
+
+use std::sync::Arc;
+
+use p3_challenger::CanObserve;
+use p3_matrix::dense::RowMajorMatrix;
+use serde::{de::DeserializeOwned, Serialize};
+
+use super::types::{
+    MultiStarkProvingKeyView, PairView, ProverViewAfterRapPhases, RapView, StarkProvingKeyView,
+};
+use crate::{
+    air_builders::symbolic::SymbolicExpressionDag,
+    config::{Com, RapPhaseSeqProvingKey, StarkGenericConfig, Val},
+    keygen::types::MultiStarkProvingKey,
+};
+
+/// Associated types needed by the prover, in the form of buffers and views,
+/// specific to a specific hardware backend.
+///
+/// Memory allocation and copying is not handled by this trait.
+pub trait ProverBackend {
+    /// Extension field degree for the challenge field `Self::Challenge` over base field `Self::Val`.
+    const CHALLENGE_EXT_DEGREE: u8;
+    // ==== Host Types ====
+    /// Base field type, on host.
+    type Val: Copy + Send + Sync + Serialize + DeserializeOwned;
+    /// Challenge field (extension field of base field), on host.
+    type Challenge: Copy + Send + Sync + Serialize + DeserializeOwned;
+    /// PCS opening proof on host (see [OpeningProver]). This should not be a reference.
+    type OpeningProof: Clone + Send + Sync + Serialize + DeserializeOwned;
+    /// Single commitment on host.
+    // Commitments are small in size and need to be transferred back to host to be included in proof.
+    type Commitment: Clone + Send + Sync + Serialize + DeserializeOwned;
+    /// Challenger to observe commitments. Sampling is left to other trait implementations.
+    /// We anticipate that the challenger largely operates on the host.
+    type Challenger: CanObserve<Self::Val> + CanObserve<Self::Commitment>;
+
+    // ==== Device Types ====
+    /// Single matrix buffer on device together with dimension metadata.
+    type MatrixView: MatrixView + Clone + Send + Sync;
+    /// Reference for the preimage of a PCS commitment on device, together with any metadata
+    /// necessary for computing opening proofs.
+    ///
+    /// For example, multiple buffers for LDE matrices, their trace domain sizes, and pointer to mixed merkle tree.
+    type PcsDataView: Clone + Send + Sync;
```

Comment:
I think it's better to define `PcsData` rather than `PcsDataView`. With `PcsData` the prover itself owns the data and can create immutable or mutable references. And the prover can even drop it on gpu and transfer it to host.

Reply:
From the host perspective, it should never own `PcsData` though? It will only ever get a reference? I feel like `PcsData` should never leave the device?

Reply:
The `PcsData` for gpu can be `DevicePointer` (the host owns the device object indirectly) and `PcsDataView` can just be `&DevicePointer`. Then you don't need to collect all the `PcsDataView`s and then invoke the `fn deallocate_pcs_datas(pdatas: Vec<PcsDataView>)` when you want to transfer the device's `PcsData` to host.